torch/_dynamo/cache_size.py

```diff
@@ -0,0 +1,152 @@
+
+There are two limits
+1) cache_size_limit
+2) accumulated_cache_size_limit
+
+
+Earlier we used to have only limit - maximum number of entries in 1 cache line
+(which is now represented by (2) above). So, why do we need two limits? Lets try
+to understand that.
+
+In general, we want our cache limit value to be a small number (e.g. 8 or even
+lower). This ensures that for frames that cause too many recompilation fall to
+eager quickly. However, there is another problem that prevents us from lowering
+the value of cache_size_limit. This is due to ID_MATCH'd guards. Today, we put
+ID_MATCH guards on nn module if there is a graph break. This means we will have
+many recompilations for the same code object because the ID_MATCH guard fails
+for different instances of the nn module. This is a common pattern in how models
+are authored. Therefore, this requires us to keep the cache_size_limit high.
+
+We resolve this by introducing these two limits. The first limit (1) limits the
+number of cache entries that have an ID_MATCH'd guard for an nn module instance.
+And, (2)nd limit becomes a safeguard mechanism to have a maximum compilations
+for a code object. One important question is - what is the limit for the code
+object that does not have any ID_MATCH guard? For such code objects, we choose
+(1) as the cache size limit.
+
+Lets take an example to understand how these limits help. Suppose, we have 16
+instances of a nn module and we ID_MATCH on the self object. Further, suppose
+the inputs to these functions have varying batch size, leading to one
+recompilation. In total, there will be 32 recompilations, and therefore 32 cache
+entries on the forward code object. In the older case when we had only 1 limit,
+our cache size limit must be >= 32 to capture all these recompilations. Now,
+suppose there is a separate function in the same program which is very dynamic
+and unsuitable for compilation. Such a function will need to undergo 32
+compilations to burst the cache and fallback to eager. These 32 recompilations
+are too many and we want to fallback for these compilation-unfriendly functions
+sooner.
+
+In the new scenario, we can have (1) cache_size_limit = 2, (2)
+accumulated_cache_size_limit = 32. This means that each ID_MATCH'd object can
+have maximum of two cache entries, and the maximum number of cache entries
+(irrespective of ID_MATCH obj) is 32. This covers the case of forward code
+object which has 32 recompilations. For the other function, the one unsuitable
+for recompilation, our limit is 2. So, we will burst the cache in just 2
+recompilations. In this manner, these 2 limits help us resolve the tension
+mentioned earlier.
+"""
+
+
+@dataclass
+class CacheSize:
+    """ """
+
+    # Total number of CacheEntry objects in the Dynamo linked list
+    num_cache_entries: int = 0
+
+    # Number of CacheEntry objects having same ID_MATCH'd objects as given frame.
+    num_cache_entries_with_same_id_matched_objs: int = 0
```

Comment:
No I'm still confused. I was expecting a Dict from id to individual count per id. How come this is all squashed into one int now?

Reply:
So, the local_names that have ID_MATCH can change from frame to frame

Consider
~~~

def fn(a, b):
    ....

In frame 1 - a has ID_MATCH - id1
In frame 2- b has ID_MATCH - id2
In frame 3 - a, b has ID_MATCH where id of a and b are same as in frame 1 and frame 2 - id1 and id2
~~~

In current state, none of these 3 frames will be tagged as a re-compilation. 

If I had the old Dict from id to individual count per id (like in earlier commit), we would have

id1 - 2
id2 - 2

But that seems wrong. This sounds like we have 4 cache entries in total, but when in reality we have only 3. 

In current state, essentially our Dict  is tuple_of_matched_ids -> count_per_tuple_of_matched_ids

(id1, ) - 1
(id2,) - 1
(id1, id2) - 1

This makes recompilation and counting cache entries consistent in definition.

Now, the reason why I don't store this new dict - its just because I don't need it. We can just keep count of cache entries that have same tuple of id_matched objects in current frame. That is enough to check both is_recompilation and exceeds_cache_size.